### PR TITLE
feat: dao voting token holder status

### PR DESCRIPTION
**Pull request by phucledien**
**What does this PR do?**

-   [x] Get token holder status for creating proposal and voting
-   [x] Update schema to use `numeric` postgres data type for `required_amount` col

**How to test**

-   Use API **GET** `{{baseURL}}/dao-voting/token-holder/status`

Query:
- `*action`: `vote` | `creat_proposal`
- `*user_id`": Discord user id, i.e: `205167514731151360`
- `proposal_id`: **required** when `action` is `vote`, i.e: `1` 
- `guild_id`:  **required** when `action` is `create_proposal`

**Media (Loom or gif)**

- Not connect wallet
CleanShot 2023-01-09 at 13 21 59@2x

- Action is `vote`
CleanShot 2023-01-09 at 13 15 29@2x
CleanShot 2023-01-09 at 13 16 38@2x
CleanShot 2023-01-09 at 13 17 14@2x

- Action is `create_proposal`
CleanShot 2023-01-09 at 13 18 49@2x
CleanShot 2023-01-09 at 13 18 19@2x

### CODE REVIEW

**Review by trkhoi — APPROVED** [review]

File: pkg/request/dao_voting.go [hunk 1]
```diff
@@ -33,3 +33,18 @@ type UpdateDaoVoteRequest struct {
...
+
+const (
+	CreateProposal DAOAction = "create_proposal"
+	Vote           DAOAction = "vote"
+)
+
+type TokenHolderStatusRequest struct {
+	Action             DAOAction `json:"action" form:"action" binding:"required,oneof=create_proposal vote"`
+	UserID             string    `json:"user_id" form:"user_id" binding:"required"`
+	ProposalID         string    `json:"proposal_id" form:"proposal_id"`
```
Comment (anhnh12) [inline pkg/request/dao_voting.go:47]: why not int64, so no need to parse

File: pkg/entities/dao_proposal.go [hunk 2]
```diff
@@ -99,3 +103,194 @@ func (e *Entity) DeleteDaoProposal(proposalId string) error {
...
+	requiredAmountBigInt, ok := new(big.Int).SetString(config.RequiredAmount, 10)
+	if !ok {
+		err = fmt.Errorf("cannot convert big int from string")
+		e.log.Fields(logger.Fields{
+			"requiredAmount": config.RequiredAmount,
+			"base":           10,
+		}).Error(err, "[entities.TokenHolderStatus] - new(big.Int).SetString failed")
+		return nil, err
+	}
+	isQualified := userBalance.Cmp(requiredAmountBigInt) != -1
```
Comment (anhnh12) [inline pkg/entities/dao_proposal.go:247]: this only true with vote_option type `nft`

File: pkg/entities/tokens.go [hunk 3]
```diff
@@ -201,3 +206,37 @@ func (e *Entity) TotalSupportedTokens(guildId string) (*response.Metric, error)
...
+		return nil, fmt.Errorf("failed to get chain by id %s: %v", chainID, err)
+	}
+
+	// TODO: Support non evm
+	client, err := ethclient.Dial(chain.RPC)
+	if err != nil {
+		return nil, fmt.Errorf("failed to connect to chain client: %v", err.Error())
+	}
+
+	contractERC20, err := erc20.NewErc20(common.HexToAddress(address), client)
```
Comment (anhnh12) [inline pkg/entities/tokens.go:228]: what about native token?

File: pkg/entities/tokens.go [hunk 4]
```diff
@@ -201,3 +206,37 @@ func (e *Entity) TotalSupportedTokens(guildId string) (*response.Metric, error)
...
+	if err != nil {
+		e.log.Errorf(err, "[erc20.NewErc20] failed to init erc20 contract")
+		return nil, fmt.Errorf("failed to init erc20 contract: %v", err.Error())
+	}
+	balanceOf := func(address string) (*big.Int, error) {
+		b, err := contractERC20.BalanceOf(nil, common.HexToAddress(address))
```
Comment (anhnh12) [inline pkg/entities/tokens.go:234]: can reuse func from `chain.go`?

**Review by anhnh12 — CHANGES_REQUESTED** [review]

File: pkg/request/dao_voting.go (hunk 1 above)
Comment (phucledien) [reply pkg/request/dao_voting.go:47]: ok ser

File: pkg/entities/tokens.go (hunk 4 above)
Comment (phucledien) [reply pkg/entities/tokens.go:234]: The version from chain.go using float64 for balance number, and the db schema is storing the required amount in numeric type (big.Int value) which I believe would mismatch the type 🤔

File: pkg/entities/dao_proposal.go [hunk 5]
```diff
@@ -99,3 +103,194 @@ func (e *Entity) DeleteDaoProposal(proposalId string) error {
...
+
+func (e *Entity) tokenHolderStatusForCreatingProposal(walletAddress string, query request.TokenHolderStatusRequest) (*response.TokenHolderStatus, error) {
+	if query.GuildID == "" {
+		return nil, errs.ErrInvalidDiscordGuildID
+	}
+	if query.GuidelineChannelID == "" {
+		return nil, errs.ErrInvalidDiscordChannelID
+	}
+
+	config, err := e.repo.GuildConfigDaoProposal.
```
[the remark this answers is not in the document]
Comment (phucledien) [reply pkg/entities/dao_proposal.go]: fixed it

**Review by anhnh12 — APPROVED** [review]

**Pull request merged**